internal/core/textlayout.rs

```diff
@@ -0,0 +1,954 @@
+//!    example arabic text can be covered by a font that has excellent arabic coverage while latin text is rendered using a different font.
+//!    Shaping boundaries are always also grapheme boundaries.
+//! 2. Then we shape the text at shaping boundaries, to determine the metrics of glyphs and glyph clusters (grapheme boundaries with the shapable)
+//! 3. Allocate graphemes into new text lines until all graphemes are consumed:
+//! 4. Loop over all graphemes:
+//!     Compute the width of the grapheme
+//!     Determine if the grapheme is produced by a white space character
+//!     If grapheme is not at break opportunity:
+//!         Add grapheme to fragment
+//!         If width of current line <= available width AND width of current line + fragment > available width:
+//!             Emit current line
+//!             Current line starts with fragment
+//!             Clear fragment
+//!         Else:
+//!              Continue
+//!     Else if break opportunity at grapheme boundary is optional OR if current is space and next is optional:
+//!         If width of current line + fragment <= available width:
+//!              Add fragment to current line
+//!              Clear fragment
+//!         Else:
+//!              Emit current line
+//!              Current line starts with fragment
+//!              Clear fragment
+//!         Add grapheme to fragment
+//!             
+//!     Else if break opportunity at grapheme boundary is mandatory:
+//!         Add fragment to current line
+//!         Emit current line
+//!         Clear fragment
+//!         Add grapheme to fragment
+//!
+
+use core::cell::RefCell;
+use core::ops::Range;
+
+use alloc::vec::Vec;
+
+use euclid::num::{One, Zero};
+
+use crate::items::{TextHorizontalAlignment, TextOverflow, TextVerticalAlignment, TextWrap};
+
+pub enum BreakOpportunity {
+    Allowed,
+    Mandatory,
+}
+
+#[cfg(feature = "unicode-linebreak")]
+mod linebreak_unicode;
+#[cfg(feature = "unicode-linebreak")]
+use linebreak_unicode::LineBreakIterator;
+
+#[cfg(not(feature = "unicode-linebreak"))]
+mod linebreak_ascii;
+#[cfg(not(feature = "unicode-linebreak"))]
+use linebreak_ascii::LineBreakIterator;
+
+pub trait TextShaper {
+    type LengthPrimitive: core::ops::Mul
+        + core::ops::Div
+        + core::ops::Add<Output = Self::LengthPrimitive>
+        + core::ops::AddAssign
+        + euclid::num::Zero
+        + euclid::num::One
+        + core::convert::From<i16>
+        + Copy;
+    type Length: euclid::num::Zero
+        + core::ops::AddAssign
+        + core::ops::Add<Output = Self::Length>
+        + core::ops::Sub<Output = Self::Length>
+        + Default
+        + Clone
+        + Copy
+        + core::cmp::PartialOrd
+        + core::ops::Mul<Self::LengthPrimitive, Output = Self::Length>
+        + core::ops::Div<Self::LengthPrimitive, Output = Self::Length>;
```

Comment:
I wonder if we really do need to be generic over the Length?
Is it because the GL backend will use f32 as a LengthPrimitive?
Don't you think we want to have the PhysicalPx incore anyway? Then it can just be `euclid::Length<Self::LengthPrimitive, PhysicalPx>`
Or would we use different length?

Reply:
> Is it because the GL backend will use f32 as a LengthPrimitive?

Yes.

> Don't you think we want to have the PhysicalPx incore anyway?

Yes, that would be nice eventually. A bit tricky because its precision will have to depend on a feature. The MCU backend will want PhysicalPx as `i16` but the GL or Qt backend as `f32`.